File: internal/cli/list.go
package cli

import (
	"os"
	"sort"
	"strings"

	"github.com/spf13/cobra"

	"github.com/adriankarlen/storyquery/internal/output"
)

func newListCmd(gf *globalFlags) *cobra.Command {
	var filter string
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List all components",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			format, err := resolveFormat(gf)
			if err != nil {
				return err
			}
			bundle, err := loadBundle(cmd.Context(), gf)
			if err != nil {
				return err
			}

			needle := strings.ToLower(strings.TrimSpace(filter))
			res := output.ListResult{Warnings: bundle.Warnings}
			for _, c := range bundle.Components.Components {
				if needle != "" &&
					!strings.Contains(strings.ToLower(c.Name), needle) &&
					!strings.Contains(strings.ToLower(c.ID), needle) {
					continue
				}
				res.Components = append(res.Components, output.SummarizeComponent(c, 0))
			}
			sort.SliceStable(res.Components, func(i, j int) bool {
				return res.Components[i].Name < res.Components[j].Name
			})
			return output.Encode(os.Stdout, format, res)
		},
	}
	cmd.Flags().StringVar(&filter, "filter", "", "case-insensitive substring filter on name/id")
	return cmd
}

File: src/commands/list.ts
import { type Renderable, encode, summarizeComponent } from "../output/index.js";
import type { ListResult } from "../output/index.js";
import { defineGlobalCommand, loadBundle, resolveFormat } from "./shared.js";

export const listCommand = defineGlobalCommand({
  meta: { name: "list", description: "List all components" },
  args: {
    filter: {
      type: "string",
      description: "case-insensitive substring filter on name/id",
      default: "",
    },
  },
  async run({ args }) {
    const format = resolveFormat(args);
    const bundle = await loadBundle(args);

    const needle = args.filter.trim().toLowerCase();
    const result: ListResult = { components: [] };
    for (const c of Object.values(bundle.components.components ?? {})) {
      const name = (c.name ?? "").toLowerCase();
      const id = (c.id ?? "").toLowerCase();
      if (needle && !name.includes(needle) && !id.includes(needle)) continue;
      result.components.push(summarizeComponent(c, 0));
    }
    result.components.sort((a, b) => a.name.localeCompare(b.name));
    if (bundle.warnings.length > 0) result.warnings = bundle.warnings;

    const payload: Renderable = { kind: "list", value: result };
    process.stdout.write(encode(format, payload));
  },
});
